<a href="https://colab.research.google.com/github/zzhining/ml_basic/blob/main/pandas_exercise_skeleton.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [10]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

np.random.seed(0)

# 과제 1
## Q1. 


<div class="alert alert-success">
[문제1]     
    
insurance.csv'파일 읽어서 df에 할당하고, 상위 5개 항목 출력하기
</div>

------------

[힌트]

pd.read_csv()

In [11]:
df = pd.read_csv('abc/insurance.csv')
df

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

In [12]:
df.head(5)

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

## Q2. 


<div class="alert alert-success">
[문제2]     
    
df의 요약(평균, 4분위수, 전체 갯수 등)정보 확인하기
</div>

------------

In [13]:
df.describe

<bound method NDFrame.describe of       age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]>

## Q3. 


<div class="alert alert-success">
[문제3]     
    
신규 컬럼(double bmi)을 생성하고, bmi의 2배가 되는 값을 할당하기
</div>

------------

In [14]:

bmi = df.iloc[:,2:3]

In [15]:
df.insert(loc=7, column='double bmi',value=bmi*2)


In [16]:
df

age     sex     bmi  children smoker     region      charges  double bmi
0      19  female  27.900         0    yes  southwest  16884.92400       55.80
1      18    male  33.770         1     no  southeast   1725.55230       67.54
2      28    male  33.000         3     no  southeast   4449.46200       66.00
3      33    male  22.705         0     no  northwest  21984.47061       45.41
4      32    male  28.880         0     no  northwest   3866.85520       57.76
...   ...     ...     ...       ...    ...        ...          ...         ...
1333   50    male  30.970         3     no  northwest  10600.54830       61.94
1334   18  female  31.920         0     no  northeast   2205.98080       63.84
1335   18  female  36.850         0     no  southeast   1629.83350       73.70
1336   21  female  25.800         0     no  southwest   2007.94500       51.60
1337   61  female  29.070         0    yes  northwest  29141.36030       58.14

[1338 rows x 8 columns]

## Q4. 


<div class="alert alert-success">
[문제4]     
    
신규 컬럼(debt)을 생성하고, 0으로 채우기
</div>

------------

In [17]:
df.insert(loc=8,column='dabt',value=0)

In [18]:
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double bmi  dabt  
0          55.80     0  
1          67.54     0  
2          66.00     0  
3          45.41     0  
4          57.76     0  
...          ...   ...  
1333       61.94     0  
1334       63.84     0  
1335       73.70     0  
1336       51.60     0  
1337       58.14     0  

[1338 rows x 9 columns]

## Q5. 


<div class="alert alert-success">
[문제5]     
    
debt에 0부터 값이 증가하도록 채우기
</div>

------------

In [19]:
df.insert(loc=8,column='debt',value=len(df))

In [20]:
len(df)

1338

In [21]:
df = df.drop(columns='dabt')


In [22]:
ii = []
for i in range(len(df)):
    ii.append(i)
    
    

df['debt'] = ii

### 문제 5번 오답 필요

In [23]:
d2['debt'] = [i+1 for i in range(len(df))] #리스트 컴프리헨션 이용해라 



NameError: name 'd2' is not defined

In [ ]:
d2

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double bmi  debt  
0          55.80     1  
1          67.54     2  
2          66.00     3  
3          45.41     4  
4          57.76     5  
...          ...   ...  
1333       61.94  1334  
1334       63.84  1335  
1335       73.70  1336  
1336       51.60  1337  
1337       58.14  1338  

[1338 rows x 9 columns]

In [ ]:
d2 = df.copy()

## Q6. 


<div class="alert alert-success">
[문제6]     
    
1번 인덱스의 정보 출력
</div>

------------

[힌트]
iloc



age                  18  
sex                male  
bmi               33.77  
children              1  
smoker               no  
region        southeast  
charges         1725.55  
double bmi        67.54  
debt                  1  

In [ ]:
df.iloc[1]

age                  18
sex                male
bmi               33.77
children              1
smoker               no
region        southeast
charges       1725.5523
double bmi        67.54
debt                  1
Name: 1, dtype: object

## Q7. 


<div class="alert alert-success">
[문제7]     
    
'double bmi'열 삭제하기
</div>

In [ ]:
df = df.drop(columns='double bmi')


In [ ]:
df

age     sex     bmi  children smoker     region      charges  debt
0      19  female  27.900         0    yes  southwest  16884.92400     0
1      18    male  33.770         1     no  southeast   1725.55230     1
2      28    male  33.000         3     no  southeast   4449.46200     2
3      33    male  22.705         0     no  northwest  21984.47061     3
4      32    male  28.880         0     no  northwest   3866.85520     4
...   ...     ...     ...       ...    ...        ...          ...   ...
1333   50    male  30.970         3     no  northwest  10600.54830  1333
1334   18  female  31.920         0     no  northeast   2205.98080  1334
1335   18  female  36.850         0     no  southeast   1629.83350  1335
1336   21  female  25.800         0     no  southwest   2007.94500  1336
1337   61  female  29.070         0    yes  northwest  29141.36030  1337

[1338 rows x 8 columns]

## Q8. 


<div class="alert alert-success">
[문제8]     
    
 'age'열, 'sex'열만 출력하기
</div>

In [24]:
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double bmi  debt  
0          55.80     0  
1          67.54     1  
2          66.00     2  
3          45.41     3  
4          57.76     4  
...          ...   ...  
1333       61.94  1333  
1334       63.84  1334  
1335       73.70  1335  
1336       51.60  1336  
1337       58.14  1337  

[1338 rows x 9 columns]

In [ ]:
df.iloc[:,0:2]

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

In [ ]:
# 무조건 인데스자리에 값이 있어야 함 
df.loc[:,['age','sex']]

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

In [26]:
df.iloc[:,:2]

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

# 과제2
## Q9.
<div class="alert alert-success">
[문제9]     
    
나이를 1살씩 증가시키기
</div>

In [ ]:
df

age     sex     bmi  children smoker     region      charges  debt
0      19  female  27.900         0    yes  southwest  16884.92400     0
1      18    male  33.770         1     no  southeast   1725.55230     1
2      28    male  33.000         3     no  southeast   4449.46200     2
3      33    male  22.705         0     no  northwest  21984.47061     3
4      32    male  28.880         0     no  northwest   3866.85520     4
...   ...     ...     ...       ...    ...        ...          ...   ...
1333   50    male  30.970         3     no  northwest  10600.54830  1333
1334   18  female  31.920         0     no  northeast   2205.98080  1334
1335   18  female  36.850         0     no  southeast   1629.83350  1335
1336   21  female  25.800         0     no  southwest   2007.94500  1336
1337   61  female  29.070         0    yes  northwest  29141.36030  1337

[1338 rows x 8 columns]

In [27]:
df['age'] = df.iloc[:,0]+1

In [ ]:
df

age     sex     bmi  children smoker     region      charges  debt
0      20  female  27.900         0    yes  southwest  16884.92400     0
1      19    male  33.770         1     no  southeast   1725.55230     1
2      29    male  33.000         3     no  southeast   4449.46200     2
3      34    male  22.705         0     no  northwest  21984.47061     3
4      33    male  28.880         0     no  northwest   3866.85520     4
...   ...     ...     ...       ...    ...        ...          ...   ...
1333   51    male  30.970         3     no  northwest  10600.54830  1333
1334   19  female  31.920         0     no  northeast   2205.98080  1334
1335   19  female  36.850         0     no  southeast   1629.83350  1335
1336   22  female  25.800         0     no  southwest   2007.94500  1336
1337   62  female  29.070         0    yes  northwest  29141.36030  1337

[1338 rows x 8 columns]

## Q10.
<div class="alert alert-success">
[문제10]     
    
charges의 표시형식을 소수점 아래 2자리로 바꾸기
</div>

[힌트]

함수 생성, 적용(apply/map)

In [ ]:
a= round(df.iloc[:,6],2)

0       16884.92
1        1725.55
2        4449.46
3       21984.47
4        3866.86
          ...   
1333    10600.55
1334     2205.98
1335     1629.83
1336     2007.94
1337    29141.36
Name: charges, Length: 1338, dtype: float64

In [ ]:
df['charges'] = df.iloc[:,6].round(2)

In [ ]:
df

age     sex     bmi  children smoker     region   charges  debt
0      20  female  27.900         0    yes  southwest  16884.92     0
1      19    male  33.770         1     no  southeast   1725.55     1
2      29    male  33.000         3     no  southeast   4449.46     2
3      34    male  22.705         0     no  northwest  21984.47     3
4      33    male  28.880         0     no  northwest   3866.86     4
...   ...     ...     ...       ...    ...        ...       ...   ...
1333   51    male  30.970         3     no  northwest  10600.55  1333
1334   19  female  31.920         0     no  northeast   2205.98  1334
1335   19  female  36.850         0     no  southeast   1629.83  1335
1336   22  female  25.800         0     no  southwest   2007.94  1336
1337   62  female  29.070         0    yes  northwest  29141.36  1337

[1338 rows x 8 columns]

## Q11.
<div class="alert alert-success">
[문제11]     
    
컬럼 정렬(알파벳 순서대로) : age, bmi, charges... 순으로 정렬
</div>

[Hint]
sort_index()


In [ ]:
df

NameError: name 'df' is not defined

## Q12.
<div class="alert alert-success">
[문제12]     
    
값 정렬(age) : 나이 순으로 정렬
</div>

[Hint]
sort_values


## Q13.
<div class="alert alert-success">
[문제13]     
    
나이('age') 평균구하기
</div>

[Hint]
mean

## Q14.
<div class="alert alert-success">
[문제14]     
    
보험료('charges')와 가장 상관관계가 높은 변수는?
</div>

[Hint]
 - 공분산: df.cov()
 - 상관계수: df.corr()

# 과제 3
## Q15.
<div class="alert alert-success">
[문제15]     
    
 남자, 여자가 각각 몇명있는지 구하기
</div>

[Hint]
pd.value_counts


## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가 3명인 데이터 출력하기
</div>

## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가  5명 이상인 데이터 출력하기
</div>

## Q17.
<div class="alert alert-success">
[문제17]     
    
 bmi가  17 미만인 데이터 출력하기
</div>

# 과제 4 
## Q18.
<div class="alert alert-success">
[문제18]     
    
성별에 따른 보험료 평균
</div>

## Q19.
<div class="alert alert-success">
[문제19]     
    
흡연여부에 따른 보험료 평균
</div>

## Q20.
<div class="alert alert-success">
[문제20]     
    
지역별 보험료 평균
</div>

## Q21.
<div class="alert alert-success">
[문제21]     
    
흡연여부 + 지역별 데이터 평균
</div>

## Q22.
<div class="alert alert-success">
[문제22]     
    
 흡연여부 + 지역별 데이터 평균에서 'charges' 만 츨력하기
</div>

## Q23.
<div class="alert alert-success">
[문제23]     
    
 흡연여부 + 지역별 데이터 크기(size)
</div>

## Q24.
<div class="alert alert-success">
[문제24]     
    
smoker 여부에 따라, group_val 컬럼에 '흡연자', '비흡연자' 표시하기
</div>